### DISCUSSION TITLE
How to present pdf with nuxt 3? (500-error when reloading page with pdf)

### DISCUSSION THREAD

**Discussion by srbolle** [nuxt/nuxt#16230] [Questions] [answered]
### Environment

    Operating System: linux (ubuntu 22.04.1)
    Node Version: v18.12.1
    Nuxt Version: 3.0.0
    Package Manager: npm 8.19.2
    Builder: vite

### Objective

    Present pdf-files using nuxt 3

**See enclosed zip-file for all project-files.**

### Problem description

I have made a simple component to present a pdf-file. I am currently using the module "vue3-pdf-app". When my nuxt component is just created, the pdf is shown in my webbrowser (running 'npm run dev' locally). However, if I reload the page I get the following error:

**500** document is not defined

at ./node_modules/vue3-pdf-app/dist/vue3-pdf-app.umd.js:26355:12
at ./node_modules/vue3-pdf-app/dist/vue3-pdf-app.umd.js:27930:12
at webpackUniversalModuleDefinition (./node_modules/vue3-pdf-app/dist/vue3-pdf-app.umd.js:3:20)
at Object. (./node_modules/vue3-pdf-app/dist/vue3-pdf-app.umd.js:10:3)
at Module._compile (node:internal/modules/cjs/loader:1159:14)
at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
at Module.load (node:internal/modules/cjs/loader:1037:32)
at Module._load (node:internal/modules/cjs/loader:878:12)
at ModuleWrap. (node:internal/modules/esm/translators:169:29)

### Some more details

When running the local server (project files included in the enclosed zip-file), I get the above 500-error as soon as I open browser at localhost:3000

If I in pages/index.vue replace 
`<my-container />`
 with 
`<test-component />`
 and then again insert
`<my-container />`
 
the component is shown correctly with the pdf-file. But as soon as I reload the page or restart the server, I get the 500-error.

I wonder if I define the component MyContainer incorrectly, or if there may be a bug.
I've also tried to use a different pdf-library with the same result.

[nuxt3-with-pdf.zip](https://github.com/nuxt/framework/files/10123078/nuxt3-with-pdf.zip)

**Comment by sewalsh** [comment, marked as answer]
Hey @srbolle, vue3-pdf-app is trying to access `Document` which isn't available server-side (i.e. first page load).

Here's a working sandbox with the package declared as a client-side plugin.

https://codesandbox.io/s/swqsut?file=/components/MyContainer.vue

**Reply by srbolle** [reply]
Thank you @sewalsh for a solution that works beautifully!

A note to my future self and others that may come back to this thread:

First the component presenting the pdf-file should be enveloped by ClientOnly, like this:

```html
<template>
    <div>
        <ClientOnly>
            <vue-pdf-app
                style="height: 100vh"
                pdf="<url-to-pdf-file>">
             </vue-pdf-app>
         </ClientOnly>
    </div>
</template>
```

Secondly, the plugins-directory in your nuxt-project should contain a file in the format "plugins/some-name.client.ts" (e.g. "vue3-pdf-app.client.ts"). Please note the word "client" towards the end of the filename. It should not be "clients" or anything else, unless you want it to be run on the server, in which case it should have the suffix ".server". (See Nuxt Docs -> Guide -> Directory Structure --> plugins).

Here is @sewalsh solution from above for plugins/vue3-pdf-app.client.ts:

```javascript
import { defineNuxtPlugin } from '#app'
import VuePdfApp from 'vue3-pdf-app';
import 'vue3-pdf-app/dist/icons/main.css';

export default defineNuxtPlugin((nuxtApp) => {
    nuxtApp.vueApp.component('VuePdfApp', VuePdfApp);
})
```
Thanks again @sewalsh for invaluable help!
